docs/design/2018-09-04-histograms-in-plan.md

```diff
@@ -0,0 +1,144 @@
+# Proposal: Maintain statistics in `Plan`
+
+- Author:     [Yiding CUI](https://github.com/winoros)
+- Last updated:  2018-09-04
+
+## Abstract
+
+This proposal proposes to maintain the histogram in `Plan`’s statistics information.
+
+
+## Background
+
+Currently, TiDB only uses statistics when deciding which physical scan method a table should use. And TiDB only stores simple statistics in the plan structure. But when deciding the join order and considering some other optimization rules, we need more detailed statistics.
+
+So we need to maintain the statistics in the plan structure to get sufficient statistics information to do optimizations.
+
+## Proposal
+
+The new `statsInfo` of `plan` should be something like the following structure:
+
+```go
+type statsInfo struct {
+	// This two attributes are already used in current statsInfo.
+	count float64
+	// The number of distinct value of each column.
+	ndv []float64
+
+	// HistColl collects histograms of columns.
+	histColl statistics.HistColl
+}
+```
+
+This struct will be maintained when we call `deriveStats`.
+
+### How to maintain it.
+
+We maintain the histogram in `Projection`, `Selection`, `Join`, `Aggregation`, `Sort`, `Limit` and `DataSource` operators.
+
+#### `Selection`
+Just use it to calculate the selectivity.
+
+And we can get the range information of each column and index involved in filters. Then we can use the range information to cut the histogram bucket of these columns.
+
+For column/index not matched in the filter. The `NDV` and `totalCnt` of each bucket can just multiply the total selectivity of filters(based on independent assumption).
+
+#### `Projection`
+Change the reflection of the map we maintained.
```

Comment:
what does this mean?

Reply:
Before projection, we have column `a`. The projection projects a onto new column c. Then we would update the `HistColl.Columns`. So that we can use `c.UniqiueID` find column `a`'s histogram.